# Figures

This repository houses all scripts used to generate figures. 


## BoxAndWhiskersPlots_LinearModelSlopes_MetricsVsAge.py

**This is the primary script for plotting and was used to generate all amino acid slopes**

This is a behemoth of a script that can do a *lot* and is replacing several disjoint scripts. 

Author : Sara Willis
Date   : March 20, 2019


### Dependencies

#### Python


The user will need Python3 to run this script with the following non-standard Python dependencies:
 
   * [matplotlib](https://anaconda.org/conda-forge/matplotlib)

   * [mysql.connector](https://anaconda.org/anaconda/mysql-connector-python)

#### MySQL

The user will  need access to the relevant MySQL database

#### R


The user will need R installed and will need to specify the command that calls RScript. The command R will not work
to run the R script that is generated by this program


### About

This script can either be used to generate figures or can create a csv file with the slope(s) and standard error(s) of the 
linear model metric~time

##### Figures
This script is designed to generate box and whiskers plots either for full genes or pfam domains metrics vs. time.
This is done by collecting all proteins or domains, sorting them either by a homology group ID 
or by pfam ID, taking the average of all data points in a particular grouping, and creating a
data frame for all homology groups with their age. The data are either then normalized using a box-cox
transformation or not (this depends on the user input) and a linear regression is performed.
The linear regression and a loess regression are plotted as an overlay on the box and whisker plot

##### Slopes File
If the user wants to create a csv file that contains the slope(s) and standard error(s) of a particular metric vs time, they can
us the input option "slopes". For things like ISD, Aggregation Propensity, etc... this will only generate a single row, however, 
if the user is looking at amino acid composition for multiple amino acids, the slopes for all amino acids analyze will be stored
in the same CSV file.

This script allows the user to enter various options to customize the dataset they are interested in.
They may choose the kingdom, the transmembrane status, domains vs. proteins, etc. that they want to analyze.
All options are listed for the user below.

*Note: order is not important for the command-line arguments*


### Usage


```
python3 ScriptName.py pfam|fullgene metric=<x> [test] [speciesuid=<n>] [kingdom=<x>] [transformed] [transmembrane=<x>] [xlim=xmin,xmax] [ylim=ymin,ymax] [aa=<x>] [slopes]
```



### Options



pfam|fullgene    :  (required) pfam or fullgene. Specifies which dataset used to generate slopes
                       
metric=<x>       :  (required) Specifies which metric will be plotted
                                                  
test             :  if included, only extracts first 1000 columns from each data table. For use with testing and debugging

speciesuid=n     :  Amino acid slopes for species with species UID n (found in SpeciesList MySQL table)
                
kingdom=x        :  Amino acid slopes for designated kingdom 
                                                          
transformed      :  Uses Box-Cox transform to normalize dataset before computing slope                                  
                    By default, if this argument is not included, the data will not be transformed
                     
transmembrane=x  :  When not included as an argument, all data points are included.                                     
                    When x=true, only includes transmembrane proteins/domains                                           
                    When x=false, only includes proteins/domains predicted to not be transmembrane
                     
aa=<x>           :  Chooses the amino acid(s) to be analyze. For all amino acids, use "all" or leave blank              
                    Otherwise use a comma-delimited string to select a subset. e.g. aa=V,L,F,M
                         
xlim=xmin,xmax   :  configures the x limits of the R plot. Must include both xmin and xmax as comma-delimited values to run
                                                             
ylim=ymin,ymax   :  configures the y limits of the R plot. Must include both ymin and ymax as comma-delimited values to run

                                                            
-------------

Domain Options   :  ISD, DensityOfAPRs, DensityOfAAsInAPRs, Clustering_Trunc, Clustering_AllFrames, Length, AAComp
     
Kingdom Options  :  invertebrate, vertebrate, plant, fungi

                                                            
Metric Options   :  ISD, DensityOfAPRs, DensityOfAAsInAPRs, Length, AAComp

                                            
### Data Transformation

Because Box-Cox transforms require values to be greater than zero, if the data are transformed, the value 0.5
is added to each data point. The reason this is required is because in some cases domains or proteins may be 
too short to include all amino acids and so sometimes the percent composition for a particular domain/protein
homology group is zero. If a particular metric is dependent on the length of the protein (i.e. Density of APRs
or AA Composition) 0.5 is added to the numerator. Because the extracted metrics have all been divided by the
protein/domain length prior to extraction, this is achieved by adding 0.5\length to the metric




## PlotPfamDataSet_FourSpecies_BoxAndWhiskers.py

Author : Sara Willis
Date   :Â February 1, 2019

**This program needs to be updated to incorporate various changes to bring it into line with the script BoxAndWhiskersPlots_LinearModelSlopes_MetricsVsAge.py**

#### TO RUN: 
```
python3 <filename.py> --Metric
```

Choose metrics option from below or type --options

The purpose of this script is to plot the means and standard errors of various protein domain metrics vs. the age of that domain. This script is intended to plot four species in a single figure in a 2x2 grid. The default species are Mus Musculus, Drosophila Melanogaster, Saccharomyces Cerevisiae, and Arabidopsis Thaliana. The species are specified by the user in the **User Data** section of the script as a list. This can be modified so the user can plot whichever species they want. To select from a list of species, see SpeciesList in the MySQL data table PFAMphylostratigraphy.

There are a variety of metrics that can be plotted. The metric is specified by the user as a command-line argument. They are the following:
   1) --ISD : The mean ISD predicted by IUPred 2
   2) --DensityOfAPRs : The number of aggregation-prone regions divided by the length of the protein calculated using the output of TANGO
   3) --DensityOfAAsInAPRs : The number of amino acids located in aggregation-prone regions divided by the length of the protein, calculated using the output of TANGO
   4) --Clustering_Trunc : The normalized index of dispersion using FILVM as the most hydrophobic amino acids and using one window
   5) --Clustering_AllFrames: The normalized index of dispersion using FILVM as the most hydrophobic amino acids, calculated using all possible windows
   6) --Length : The length of the amino acid sequence



## SpeciesSpecific_PlotAverageMetricVsAge_PfamOnly.py


**This is currently an outdated script. A better option would be to use BoxAndWhiskersPlots_LinearModelSlopes_MetricsVsAge.py for single-species analyses**

Author : Sara Willis
Date   : February 4, 2019

#### TO RUN:
```
python3 <filename.py> --Metric [--y_min=VALUE --y_max=VALUE]
```
The purpose of this script is to generate scatterplots of the means and standard errors of a user-specified metric vs. Age for pfam domains across individual species. 

The user should specify the metric that should be plotted in the command line using the format specified under TO RUN. The options are the following:

   1) --ISD : The mean ISD predicted by IUPred 2
   2) --DensityOfAPRs : The number of aggregation-prone regions divided by the length of the protein calculated using the output of TANGO
   3) --DensityOfAAsInAPRs : The number of amino acids located in aggregation-prone regions divided by the length of the protein, calculated using the output of TANGO
   4) --Clustering_Trunc : The normalized index of dispersion using FILVM as the most hydrophobic amino acids and using one window
   5) --Clustering_AllFrames: The normalized index of dispersion using FILVM as the most hydrophobic amino acids, calculated using all possible windows
   6) --Length : The length of the domain

The user also has the option to make the various plots have uniform axes by specifying a y_min and y_max value for each figure. These are optional arguments. If none are specified, matplotlib uses the default settings. 
